allennlp/data/dataset.py

```diff
@@ -170,3 +186,40 @@ def as_tensor_dict(self,
         for field_name, field_tensor_list in field_tensors.items():
             final_fields[field_name] = field_classes[field_name].batch_tensors(field_tensor_list)
         return final_fields
+
+    @overrides
+    def __iter__(self) -> Iterator[Instance]:
+        return iter(self.instances)
+
+
+
+class LazyDataset(Dataset):
+    """
+    A Dataset that contains a way of generating instances, rather than a
+    concrete list of them.
+
+    Parameters
+    ----------
+    instance_generator: ``Callable[[], Iterator[Instance]])``
+        This function should be callable multiple times, and each time it should
+        return an iterator that ranges over all instances.
+    """
+    def __init__(self,
+                 instance_generator: Callable[[], Iterator[Instance]]) -> None:
+        super().__init__()
+        self.generator = instance_generator
+        self.vocab: Vocabulary = None
+
+    @overrides
+    def index_instances(self, vocab: Vocabulary) -> None:
+        self.vocab = vocab
+
+    @overrides
+    def __iter__(self) -> Iterator[Instance]:
+        if self.vocab is None:
+            logger.warning("iterating over an unindexed dataset")
```

Comment:
I don't think this is necessary - if you try to use an un-indexed `Instance` to make a tensor, it'll crash.  This just adds a warning for no reason when you're building the vocab.  I was wrong to suggest it before.